server-timing/class-perflab-server-timing.php

```diff
@@ -0,0 +1,255 @@
+		 *
+		 * This action is the last possible point to register a Server-Timing metric.
+		 *
+		 * @since n.e.x.t
+		 */
+		do_action( 'perflab_server_timing_send_header' );
+
+		$header_value = $this->get_header_value();
+		if ( ! $header_value ) {
+			return;
+		}
+
+		header( sprintf( 'Server-Timing: %s', $header_value ), false );
+	}
+
+	/**
+	 * Gets the value for the Server-Timing header.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return string The Server-Timing header value.
+	 */
+	public function get_header_value() {
+		// Get all metric header values, as long as the current user has access to the metric.
+		$metric_header_values = array_filter(
+			array_map(
+				function( Perflab_Server_Timing_Metric $metric ) {
+					$value = $metric->get_value();
+					if ( null === $value ) {
+						return null;
+					}
+
+					// Check the registered capability here to ensure no metric without access is exposed.
+					if ( ! current_user_can( $this->registered_metrics_data[ $metric->get_slug() ]['access_cap'] ) ) {
+						return null;
+					}
+
+					return $this->format_metric_header_value( $metric );
+				},
+				$this->registered_metrics
+			),
+			function( $value ) {
+				return null !== $value;
+			}
+		);
+
+		return implode( ', ', $metric_header_values );
+	}
+
+	/**
+	 * Returns whether an output buffer should be used to gather Server-Timing metrics during template rendering.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return bool True if an output buffer should be used, false otherwise.
+	 */
+	public function use_output_buffer() {
+		/**
+		 * Filters whether an output buffer should be used to be able to gather additional Server-Timing metrics.
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @param bool $use_output_buffer Whether to use an output buffer.
+		 */
+		return apply_filters( 'perflab_server_timing_use_output_buffer', false );
+	}
+
+	/**
+	 * Hook callback for the 'template_include' filter.
+	 *
+	 * This effectively initializes the class to send the Server-Timing header at the right point.
+	 *
+	 * This method is solely intended for internal use within WordPress.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param mixed $passthrough Optional. Filter value. Default null.
+	 * @return mixed Unmodified value of $passthrough.
+	 */
+	public function on_template_include( $passthrough = null ) {
+		if ( ! $this->use_output_buffer() ) {
+			$this->send_header();
+			return $passthrough;
+		}
+
+		ob_start();
+		add_action(
+			'shutdown',
+			function() {
+				$output = ob_get_clean();
+				$this->send_header();
+				echo $output;
+			},
+			-1000
+		);
+		return $passthrough;
+	}
+
+	/**
+	 * Formats the header segment for a single metric.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Perflab_Server_Timing_Metric $metric The metric to format.
+	 * @return string Segment for the Server-Timing header.
+	 */
+	private function format_metric_header_value( Perflab_Server_Timing_Metric $metric ) {
+		$value = $metric->get_value();
+		if ( is_float( $value ) ) {
```

Comment:
should this have another check to make sure the value is a number (ie set at all)

Reply:
@adamsilverstein Such a check is already present in `get_header_value()` method, which ensures this won't be called if no value is set. But good point, it's probably a good idea to move that in here.